### ISSUE TITLE
Use the GPU for texture compression instead of the CPU

### ISSUE THREAD

**Issue by jonnyawsom3** [Yellow-Dog-Man/Resonite-Issues#2488]
### Is your feature request related to a problem? Please describe.

Resonite uses AMD's [Compressonator](https://github.com/GPUOpen-Tools/compressonator) under the hood to compress textures into GPU formats, but currently it only uses the CPU.
When setting local high resolution textures to BC7, this results in many minutes of 100% CPU usage, making the system barely responsive.

### Describe the solution you'd like

The library supports GPU encoding with a fallback to CPU (in the command line tool at least...) meaning compatibility shouldn't change.
In testing, the GPU encoding was 7x faster than CPU with equal or higher quality, so it should be a healthy trade-off while also lessening the CPU load of Resonite.

### Describe alternatives you've considered

Not using high quality formats such as BC7, or making sure to first save textures in the inventory, to offload to the asset variant system.

### Requesters

jonnyawsom3 draycethevoidangel

**Comment by Frooxius** [comment]
We could try this as an option maybe.

My concern is that this will cause severe stuttering, particularly when running in VR, since GPU's aren't nearly as good as CPU's at multitasking, so this could end up in a lot worse user experience.

With CPU encoding, we can better constrain it to fewer threads to reduce impact on the framerate.

**Comment by gameboycjp** [comment]
Pretty sure this is a dupe of #1983, it's something I've been paying close attention to. The stuttering is an issue on CPU too because it gets hammered so hard anyways.

**Comment by jonnyawsom3** [comment]
I was going to reference that issue but couldn't remember the name or number.

Currently most users are at a CPU bottleneck anyway, so it could be less of an issue after the performance update.
Ideally we could have an option like Froox said between CPU and GPU, with Thread Count to avoid the lockup (Maybe 'Auto' for Physical Cores only?). Then people can change it depending on their own hardware

**Comment by Frooxius** [comment]
The CPU bottlenecking doesn't necessarily pose a problem here - usually users being CPU bottlenecked doesn't mean that CPU is fully at 100 % - there can be free cores and free time on cores, which can be used at low priority, which will generally not interfere much.

In comparison GPU doesn't really allow for multi-tasking the way CPU's do, so it can be much harder to make it not impact user's smooth experience, even with they are CPU bottlenecked.

**Comment by gameboycjp** [comment]
The CPU gets hardlocked at 100% on every thread for minutes at a time, and, at least in my experience, causes unbearable stuttering in steamvrs compositor, and makes the rest of the system fairly unresponsive.

**Comment by Frooxius** [comment]
Yes, that might be the current behavior, but my point is - that can be changed relatively easily - we constrain the computation to smaller number of threads and give it a lower priority, so it doesn't take away from other higher priority processes.

GPU's generally don't have such concepts. It will block everything until it's done, which makes it much harder for GPU accelerated processes to not interfere as much.

**Comment by kulzae** [comment]
I haven't experienced my system locking when flipping through compression settings. Perhaps the locking is related to the same issue causing https://github.com/Yellow-Dog-Man/Resonite-Issues/issues/1983 ? As having the system lock up is something I haven't seen untill somewhat recently

**Comment by FlameSoulis** [comment]
This may introduce another related issue, but one of Resonite's greatest issues is the immediate application of a selection, even if merely viewing said option is a requirement. Maybe a delay or 'apply' button may be needed for those conditions. I'll do some Git surfing later on to see if such a ticket exists.

**Comment by CybrixTheProot** [comment]
> We could try this as an option maybe.
> 
> My concern is that this will cause severe stuttering, particularly when running in VR, since GPU's aren't nearly as good as CPU's at multitasking, so this could end up in a lot worse user experience.
> 
> With CPU encoding, we can better constrain it to fewer threads to reduce impact on the framerate.

The whole game stutters. I don't think stuttering during a stutter will effect the user experience that much.

**Comment by Frooxius** [comment]
> > We could try this as an option maybe.
> > My concern is that this will cause severe stuttering, particularly when running in VR, since GPU's aren't nearly as good as CPU's at multitasking, so this could end up in a lot worse user experience.
> > With CPU encoding, we can better constrain it to fewer threads to reduce impact on the framerate.
> 
> The whole game stutters. I don't think stuttering during a stutter will effect the user experience that much.

Read the whole comment (or the whole chain) please, that's something that's already addressed.

To reiterate - with CPU encoding we can constrain things more easily so it does **not** stutter. By switching to GPU encoding we might loose that ability to remove/reduce the stuttering.

**Comment by Zyro1331** [comment]
Someone in the Discord mentioned the idea of applying this to Headless servers instead, as that use-case would make the GPU much less of a priority. It could help keep some things free for the main process, accelerate the compression speeds, and not be much of a disruption due to the GPU not being used actively for rendering. 

I'm not entirely sure how often the Headless is relied upon to handle texture compression, but this would be an interesting consideration.

**Comment by Frooxius** [comment]
There's to main problems with that:
- Headlesses often run on hardware that doesn't necessarily have (good) GPU acceleration necessary for this
- Headless only compresses the textures in some cases - ones that would not help here. If you're having stutters, then it means that headless is already NOT compressing the textures - you are. Meaning having headless use GPU to compress would do nothing to help.

**Comment by ProbablePrime** [comment]
I will do some experiments though.

**Comment by jonnyawsom3** [comment]
> `this results in many minutes of 100% CPU usage` is [#1983](https://github.com/Yellow-Dog-Man/Resonite-Issues/issues/1983) and fixing it as described here is lowering thread counts and quality.

I had tested BC6H and knew GPU compression isn't available, and so couldn't fix the other open issue. My complaint about the CPU time was originally focused on it being multiple minutes, but freezing the game was another major issue that took presidence, and made Froox's counter-argument about GPU compression halting the game feel more acceptable. 

> The issue would have better been phrased as "BC7 is slow and produces poor results" with additional context being "What about trying GPU?"

When this issue was opened a year ago, I wasn't aware of the quality issues with BC7 (I assume you saw my post on the modding Discord last night?), so it seemed like GPU compression would only improve BC7's speed.

I apologise for the missed overlap. I was trying to avoid saying this would fix the BC6H issue, when I already knew it doesn't work for that format. When I'm on my computer shortly, I'll make another post about the BC7 quality issues and how to fix them.

**Comment by jonnyawsom3** [comment]
> MR.COMPRESSONATOR!!!

Oh believe me, compared to Nvidia's texture compressor, AMDs one is... Quite special to work with...

Having caught up on the devlog channel, I wonder if the 0.6 Quality was actually a mistake from trying to increase the quality from 0.05 to 0.06. Many image/texture related problems in Resonite seem to be due to default values being used or parameters misunderstood, like https://github.com/Yellow-Dog-Man/FreeImage/pull/17, which which would both speed up screenshots and improve quality/filesize. The same is happening for PNG, but I can't find the right code to change, and I'm getting distracted. :P

The unused max_threads was a big one, but the default quality setting for BC7 makes it almost unusable.
Timings are using 6 threads as per [max_threads](https://github.com/Yellow-Dog-Man/Resonite-Issues/issues/1983#issuecomment-2983529301) on a 2K texture.
AMD BC1 | Current Resonite (10 Seconds) | AMD GPU (2 Seconds)
|-----------|----------------------------------|----------------------------|
![Image](https://github.com/user-attachments/assets/2afc7a56-0c12-42ea-a329-f84d13440468) | ![Image](https://github.com/user-attachments/assets/11128f6c-44ac-4b85-9ee3-01a4bb8b4fca) | ![Image](https://github.com/user-attachments/assets/cbfddf55-0210-487e-b748-a1e6fd65752c)
| 4K Texture Timings: | 42 Seconds (300 Seconds of CPU time) | 4 Seconds

So should be a considerable speedup and decent quality improvement, although...

Quality 0.3 (100 Seconds) | NVIDIA GPU (1 Second)
|--------------------------|--------------------------|
![Image](https://github.com/user-attachments/assets/92814016-fb14-4fda-8ab0-2da6ff5c0d75) | ![Image](https://github.com/user-attachments/assets/24383df3-185d-4d46-9e5a-276df50938e5)

AMD's best manages to be worse than NVIDIA's fastest... I know we can't change it due to compatibility, but AMD is a decade behind... (This took me 2 hours to format properly, thanks Github)

**Comment by jonnyawsom3** [comment]
BC7 is correct, and thank you. Only the first image is BC1, which I added for comparison against the currently very poor BC7 quality, since it debatably looks better at half the memory size which clearly isn't right.

Generally it's industry standard to 'upgrade' any BC3/DXT5 textures to BC7, as it's the same memory cost but near-lossless to the source. The same goes for DXT5nm to BC5 for Normal Maps, same size but near-lossless. Currently Resonite does neither, which is why you've rarely seen them, but given the current speed and quality, it's no wonder why it hadn't been done before.

It's a can of worms digging into texture and image compression, since there's always so many variables involved, often literally. I'm happy to help you as much as I can though, since with the right tweaks, things could be significantly faster while also looking better for essentially free.

**Comment by jonnyawsom3** [comment]
Did a quick test, it's over 10x faster and slightly higher quality (SSIM, ignore PSNR).
![MR.COMPRESSONATOR](https://github.com/user-attachments/assets/f5c908b0-f027-4f3b-8320-31e4fcfdea2f)
Unfortunately, unlike Nvidia, the GPU compressor has no quality parameter, so that's as good as we can get.
